volatility3/framework/plugins/windows/mbrscan.py

```diff
@@ -0,0 +1,109 @@
+# This file is Copyright 2022 Volatility Foundation and licensed under the Volatility Software License 1.0
+# which is available at https://www.volatilityfoundation.org/license/vsl-v1.0
+#
+
+import logging
+import hashlib
+
+from volatility3.framework import constants, exceptions, interfaces, renderers, symbols
+from volatility3.framework.configuration import requirements
+from volatility3.framework.layers import scanners
+from volatility3.framework.renderers import format_hints
+from volatility3.framework.symbols import intermed
+from volatility3.framework.symbols.windows.extensions import mbr
+
+vollog = logging.getLogger(__name__)
+
+class MBRScan(interfaces.plugins.PluginInterface):
+    """ Scans for and parses potential Master Boot Records (MBRs) """
+
+    _required_framework_version = (2, 0, 1)
+    _version = (1, 0, 0)
+
+    @classmethod
+    def get_requirements(cls):
+        return [
+            requirements.ModuleRequirement(name = 'kernel', description = 'Windows kernel',
+                                           architectures = ["Intel32", "Intel64"])            
+        ]
+
+    @classmethod
+    def get_hash(cls, data:bytes) -> str:
+        return hashlib.md5(data).hexdigest()
+
+    def _generator(self):
+        kernel = self.context.modules[self.config['kernel']]
+        physical_layer_name = self.context.layers[kernel.layer_name].config.get('memory_layer', None)
+        
+        # Decide of Memory Dump Architecture
+        layer = self.context.layers[physical_layer_name]
+        architecture = "intel" if not symbols.symbol_table_is_64bit(self.context, kernel.symbol_table_name) else "intel64"
+
+        # Read in the Symbol File
+        symbol_table = intermed.IntermediateSymbolTable.create(context = self.context,
+                                                               config_path = self.config_path,
+                                                               sub_path = "windows",
+                                                               filename = "mbr",
+                                                               class_types = {
+                                                                'PARTITION_TABLE': mbr.PARTITION_TABLE,
+                                                                'PARTITION_ENTRY': mbr.PARTITION_ENTRY
+                                                               })
+
+        partition_table_object = symbol_table + constants.BANG + "PARTITION_TABLE"
+        
+        # Define Signature and Data Length
+        mbr_signature = b"\x55\xAA"
```

Comment:
This is a very small signature, are there more checks that can be put in place to ensure we don't hit random sets of 0x200 bytes that start b"\x55\xAA"?  Do MBRs always appear at index 0x0 within a page, etc?

Reply:
As you know, MBR data consists of three areas.
Among them, the **Partition Table** depends on the user's environment/configuration,
so it is not appropriate to write as a detection feature.
**Boot Code** mostly has the same value,
but it is likely to be modulated by malicious code as specified [in the article](https://volatility-labs.blogspot.com/2012/10/movp-43-recovering-master-boot-records.html). 
Considering this, relatively random sets (like False Positive) can be returned due to the short length of the **Signature**,
but I think it is the best way to reduce False Negative in memory analysis.
(Actually, the same signature is used for [Volatility 2](https://github.com/volatilityfoundation/volatility/blob/master/volatility/plugins/mbrparser.py#L282)).
```python
scanner = MBRScanner(needles = ['\x55\xaa'])
            for offset in scanner.scan(address_space):
                PARTITION_TABLE = obj.Object('PARTITION_TABLE', vm = address_space,
                               offset = offset)
```

Reply:
Yep, fair enough, I guess we'll have to see how much the false positives outweigh the false negatives.

Reply:
That's right. I'll update it based on User comments and facts that can be confirmed later.